samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,295 @@
+    private String system;
+    private String physicalName;
+    private int partition;
+
+    public String getStreamId() {
+      return streamId;
+    }
+
+    public void setStreamId(String streamId) {
+      this.streamId = streamId;
+    }
+
+    public String getSystem() {
+      return system;
+    }
+
+    public void setSystem(String system) {
+      this.system = system;
+    }
+
+    public String getPhysicalName() {
+      return physicalName;
+    }
+
+    public void setPhysicalName(String stream) {
+      this.physicalName = stream;
+    }
+
+    public int getPartition() {
+      return partition;
+    }
+
+    public void setPartition(int partition) {
+      this.partition = partition;
+    }
+  }
+
+  static final class Job {
+    private String jobName;
+    private String jobId;
+    private List<OpNode> opNodes;
+
+    public String getJobName() {
+      return jobName;
+    }
+
+    public void setJobName(String jobName) {
+      this.jobName = jobName;
+    }
+
+    public String getJobId() {
+      return jobId;
+    }
+
+    public void setJobId(String jobId) {
+      this.jobId = jobId;
+    }
+
+    public List<OpNode> getOpNodes() {
+      return opNodes;
+    }
+
+    public void setOpNodes(List<OpNode> opNodes) {
+      this.opNodes = opNodes;
+    }
+  }
+
+  static final class GraphNodes {
+    private List<Job> jobs;
+    private List<StreamNode> streams;
+
+    public List<Job> getJobs() {
+      return jobs;
+    }
+
+    public void setJobs(List<Job> jobs) {
+      this.jobs = jobs;
+    }
+
+    public List<StreamNode> getStreams() {
+      return streams;
+    }
+
+    public void setStreams(List<StreamNode> streams) {
+      this.streams = streams;
+    }
+  }
+
+  private int nextID = 0;
+  // streamId to StreamNode mapping
+  private final Map<String, StreamNode> streamNodes = new HashMap<>();
+  // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+  // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
+    StreamGraphImpl streamGraph = (StreamGraphImpl) jobGraph.getStreamGraph();
+    GraphNodes nodes = new GraphNodes();
+    List<Job> jobs = jobGraph.getJobNodes().stream()
+        .map(jobNode -> buildJob(jobNode, streamGraph, jobGraph))
+        .collect(Collectors.toList());
+    nodes.setJobs(jobs);
+    nodes.setStreams(createListAndSort(streamNodes.values()));
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    ObjectMapper mapper = new ObjectMapper();
+    mapper.writeValue(out, nodes);
+    return new String(out.toByteArray());
+  }
+
+  private Job buildJob(JobNode jobNode, StreamGraphImpl streamGraph, JobGraph jobGraph) {
```

Comment:
There's definitely something odd about the naming or modularity. 

I'd expect a method called "buildJob"  to have an implementation that ONLY has the last 5 lines here.  So like:
`

  private Job buildJob(JobNode jobNode, Map<Integer, OpNode> opNodes) {

     Job job = new Job();
     job.setJobName(jobNode.getJobName());
     job.setJobId(jobNode.getJobId());
     job.setOpNodes(createListAndSort(opNodes.values()));
     return job;
   }

`
It does what its name implies. It builds a job and does NOTHING else.